### DISCUSSION TITLE
Linux Build Stage3D Profile Capabilities: Software / Baseline Constrained

### DISCUSSION THREAD

**Discussion by phanxgames** [airsdk/Adobe-Runtime-Support#1103] [Q&A]
Hello, first I just want to quickly say that I am so excited to learn that the Linux support has been restored and I eagerly bought the professional license to start playing with it. My game's community has been asking for this for years.

Once I got past the issue with libcurl that was discussed [in this Issue](https://github.com/airsdk/Adobe-Runtime-Support/issues/94#issuecomment-875943222) , I was able to get the game to compile.  But running the game is proving to be more difficult.

The Starling/Stage3D part of the game reporting that the render mode is "software" and that the profile is "baseline_constrained" and my capability check of creating an empty 4096x4096 texture fails.

My Starling.context.driverInfo says "Software Hw_disabled=unavailable (Baseline Constrained)".

Its possible that my Linux machine is not configured properly, but in-case its not I would like to start this discussion. I noticed someone in the [Starling forum](https://forum.starling-framework.org/d/23144-how-to-build-air-game-for-linux/32) having a similar issue.

I want to add a quote from the user that goes by the name General , he shared this in his reddit post, which matches exactly what's going wrong for me here:

> This game uses starling and Starling 2.6 and 2048x2048 textures. It works well. However, when I tried to build another game, which has 2048x4096 textures, the attempt to load 4096 texture on Linux causes Starling Error 3683 "Texture is too big". If I tried to initiate Starling with Context3DProfile.BASELINE_EXTENDED as the profile parameter for Starling creation, I got Starling Fatal error right from the start. Looks like AIR update needs to enable 4k textures on Linux.

**Comment by ajwfrost** [comment]
Okay thanks, we can check this. Just to confirm - what distribution/version of Linux are you using here?

thanks

**Reply by GeneralVimes** [reply]
@phanxgames thank you for bringing this issue up! 
@ajwfrost yes, it's right, the current Linux builds are unable to load textures bigger than 2048x2048. I wrote a tutorial from my experience about building the game for Linux: https://www.reddit.com/r/as3/comments/p9vqd2/building_air_game_for_linux_working_experience/
I was building on Ubuntu 20. Both me and players who tested the game experienced no issues for the gameversion which had only 2048x2048 textures, but for the bigger textures we were getting Starling Error 3683 "Texture is too big". 

If I tried to initiate Starling with Context3DProfile.BASELINE_EXTENDED as the profile parameter I got Fatal error right from the start.

**Reply by phanxgames** [reply]
My linux operating system is currently **Ubuntu 18.04.4 LTS (GNU/Linux 5.4.0-81-generic x86_64)**
and it is a physical computer, not a VM, with a discrete nvidia video card (GeForce GT 640 with 1GB of VRAM), with the proprietary nvidia drivers installed.

### Forcing BASELINE_EXTENDED

I also get an error if I start up Starling with a forced BASELINE_EXTENDED Context3DProfile:
> [Starling] Context3D not available! Possible reasons: wrong renderMode or missing device support.

I have tried setting renderMode to both gpu and direct in my application-descriptor.xml, both have the same error.

### Auto Defaults to Software Mode

Setting Starling to initialize with the default settings, which is to auto detect the renderMode and context3Dprofile, reports that Starling starts up in "Software" mode with "Baseline Constrained". 

GeneralVimes replied to me on the Starling forum post with something that confirms that even though his game is running well for his players it may be running in "Software" mode none-the-less. For my game Software mode will not cut it, as I am pushing the limits of Starling a bit.

In summary:
I do believe that the Linux build is not detecting the graphics card and is falling back to Software mode.

**Comment by phanxgames** [comment]
@ajwfrost Should I create this into an issue?

**Reply by ajwfrost** [reply]
No real need to change the type of this as we're working on it anyway with this reference (#1103) and searches will generally show things from both 'issues' and 'discussions'..

**Reply by phanxgames** [reply]
Thank you so much. You are awesome. I really appreciate all you all are doing with Adobe AIR.

**Comment by pol2095** [comment]
Same problem using Intel HD Graphics
`Software Hw_disabled=unavailable (Baseline Constrained`)

**Comment by ajwfrost** [comment]
Hi all - it looks like we had some more work to do to ensure we can correctly configure and load the appropriate OpenGL libraries and extensions, so there's been a bit of an update... would be grateful if you could try with the below libCore.so file and see if this helps or what issues arise. I believe we're introducing another dependency (libGLEW.so) and am not completely sure how best to reflect that; ideally it would be something that's installed on your users' machines as/when necessary if they receive an AIR application, so something we could set up when building an RPM or DEB package that contains an application. For now -> may need to be a little bit manual, although it may also be that most folk already have this...

https://transfer.harman.com/message/tOTLapaCspm223dGH1AujC
Should go into the SDK under runtimes/air/linux/Adobe AIR/Versions/1.0/

thanks

**Reply by phanxgames** [reply]
Thank you for working so hard on this.

My system either didn't have libGLEW on it, or had the wrong version. When I attempted to install it by installing the packages libglew-dev or glew-utils, the only version of the .so file I could find was the libGLEW.so.2.0 installed to my system.

The AIR dependency was looking for libGLEW.so.1.13 , a very specific version, and I hunted for it and found it. I was able to get the AIR app to open, but it still has the same "Software Hw_disabled=unavailable (Baseline Constrained)"

Speaking of dependency problems, I am also having an issue with libcurl.so.4 (CURL_OPENSSL_3) - I was able to resolve this as well by finding that specific version and placing it next to the libCore.so , but as you said, this may need to be handled on the user's machine. Just thought I would note that there is issues with that dependency as well.

**Comment by pol2095** [comment]
How to install libGLEW.so.1.13 on Ubuntu 20 ?

I have this message error
`Error loading the runtime (libGLEW.so.1.13: cannot open shared object file: No such file or directory)`

I tried to install
`sudo apt-get install libglew-dev`
but only `libGLEW.so.2.1` is installed ?

thanks

**Reply by phanxgames** [reply]
I will be 100% honest with you, I am not sure either. But I did find this .deb install that adds the libGLEW1.13.so to my system:
https://launchpad.net/ubuntu/xenial/amd64/libglew1.13/1.13.0-2 

However, it still runs as "Software Hw_disabled=unavailable (Baseline  Constrained)"

**Reply by phanxgames** [reply]
It was extremely difficult to find the correct version of libGLEW. I tried to install it using the packages mentioned in the attachment (glew-utils, libglew-dev) but it installed libGLEW2.0  .
And then the AIR app wouldn't start until I got the 1.13 version in place.

**Reply by Cheeseness** [reply]
If it's helpful to know, you can use the LD_LIBRARY_PATH environment variable to tell ld (the part of a Linux OS that loads linked libraries at runtime) to look in a specific place for stuff like libGLEW. In some cases, this may be preferable to installing new system libs while testing things (since that may be disruptive to the system environment if it's also being used for other things).

You can set it using the export command, or you can set it just for the application that needs to be run like so, where : acts as a separator between search paths, and including $LD_LIBRARY_PATH makes sure that anything that's already in that environment variable is preserved.

    LD_LIBRARY_PATH=$LD_LIBRARY_PATH:./library_folder ./application

**Reply by pol2095** [reply]
I added `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:./libGLEW`
but adl still runs as "Software Hw_disabled=unavailable (Baseline Constrained)"

**Reply by ajwfrost** [reply]
Thanks for all the responses ... I think we'll need to look at how we load these libraries in, and potentially distribute some of them where licenses allow (or distribute everything with a package manager). But a bit concerning that even after working through those, you're still seeing hardware support as unavailable... We'll try adding a bunch of diagnostics into the runtime to see whereabouts it's going wrong!

thanks

**Reply by ajwfrost** [reply]
Ah - missing a step, I just realised :-( that currently the runtime is by default going to avoid any OpenGL stuff, unless there's a setting to override the GPU validation. We'll look at making this work properly but in the meantime, it may work better if you're able to create a file `/etc/adobe/mms.cfg` with the contents `OverrideGPUValidation=true`

**Reply by Cheeseness** [reply]
With a build that @phanxgames had given me previously, I needed to remove the libCurl he had included in order to get past that on Fedora 29. That error message seems like it's a generic fallback that doesn't communicate much about what's actually going on.

**Reply by ajwfrost** [reply]
Yes that message means that there is some unmet dependency in the libCore.so file .. if you go to where that's located in a terminal window, and run "ldd libCore.so | grep not" then it shows which libraries are not found. We were just looking at this and how to get it to find an appropriate curl library (as well as potentially removing the dependency on GLEW and trying to get it to detect/load the relevant GL libraries dynamically). It's a fair number of changes that are needed though :-(

The message itself is very wrong! and has showed us that (a) there's a build flag missing to tell the loading code that it's actually in 64-bit mode (quite why it wasn't set up to work it out itself via compiler settings, I don't know) and (b) there were some hidden version number codes left in a build file that still said 32.0!

Getting there, I think, step by step....

thanks

**Reply by Cheeseness** [reply]
Good call on checking ldd, although it won't be so helpful if it's an indirectly linked library that's missing (eg: libCurl was the friction point for the build I mentioned, but it was that libCurl's dependencies of specific libssl and libcrypto versions which would show up as "not found" via ldd).

lddtree would be a better fit for that case, but it doesn't respect rpath, so to get useful hierarchical info, it'd need to be run with LD_LIBRARY_PATH set to libCore's folder (plus any other stuff that's in libCore's rpath?). Grepping to just those lines has the same problem as ldd output, but we can use -z in GNU grep to just show everything and highlight the found terms with something like:

    LD_LIBRARY_PATH=Adobe\ AIR/Versions/1.0/ lddtree Adobe\ AIR/Versions/1.0/libCore.so | grep -z "not found"

**Reply by pol2095** [reply]
the old `libCore.so` work on Fedora using `Software Hw_disabled=unavailable (Baseline Constrained)`, I will try a RPM, I think yum can resolve the dependencies.

**Reply by ajwfrost** [reply]
Still with a libCurl dependency I'm afraid, but the below version should remove the GLEW dependency and should output an error code if it fails to set up a hardware rendering context...
https://transfer.harman.com/message/UA2p3HqRQJqYFvZsHhraVM
thanks

**Reply by ajwfrost** [reply]
.. actually don't bother, looks like there's an extension that's needed (at least for some of the test cases we have) which is fairly fundamental but isn't being picked up in this one...

**Reply by ajwfrost** [reply]
Okay -> please try this one!
https://transfer.harman.com/message/vKdrT91Ip0418YYOYt0oai
thanks

**Reply by pol2095** [reply]
Great job, OpenGL work correctly now on Ubuntu 20 and Fedora 34.
